UnitySDK/Assets/ML-Agents/Scripts/Grpc/RpcCommunicator.cs

```diff
@@ -217,42 +231,34 @@ void SendCommandEvent(CommandProto command)
 
         public void DecideBatch()
         {
-            if (m_CurrentAgents.Values.All(l => l.Count == 0))
+            if (!m_NeedCommunicateThisStep)
             {
                 return;
             }
-            foreach (var brainKey in m_CurrentAgents.Keys)
-            {
-                using (TimerStack.Instance.Scoped("AgentInfo.ToProto"))
-                {
-                    if (m_CurrentAgents[brainKey].Count > 0)
-                    {
-                        foreach (var agent in m_CurrentAgents[brainKey])
-                        {
-                            // Update the sensor data on the AgentInfo
-                            agent.GenerateSensorData();
-                            var agentInfoProto = agent.Info.ToAgentInfoProto();
-                            m_CurrentUnityRlOutput.AgentInfos[brainKey].Value.Add(agentInfoProto);
-                        }
+            m_NeedCommunicateThisStep = false;
 
-                    }
-                }
-            }
             SendBatchedMessageHelper();
-            foreach (var brainKey in m_CurrentAgents.Keys)
-            {
-                m_CurrentAgents[brainKey].Clear();
-            }
         }
 
         /// <summary>
         /// Sends the observations of one Agent. 
         /// </summary>
         /// <param name="brainKey">Batch Key.</param>
         /// <param name="agent">Agent info.</param>
-        public void PutObservations(string brainKey, Agent agent)
+        public void PutObservations(string brainKey, AgentInfo info, List<ISensor> sensors, Action<AgentAction> action)
```

Comment:
Not sure I like passing the AgentInfo here - it means multiple things are required for updating it, i.e. Agent sets the reward, but this sets the observations.

What about keeping the sensor data generation in Agent, but passing the list of observations to IPolicy.RequestDecision and then to here? Hmm, crap, that changes infererence though.

Reply:
Yes, I would need to only generate it when the Policy is communicating, which is bad...